contracts/Distributor.sol

```diff
@@ -0,0 +1,293 @@
+pragma solidity 0.5.7;
+pragma experimental ABIEncoderV2;
+
+import * as ERC721 from "@openzeppelin/contracts/token/ERC721/ERC721Full.sol";
+import * as ERC721Enumerable from "@openzeppelin/contracts/token/ERC721/ERC721Enumerable.sol";
+import * as IERC20 from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import * as Ownable from "@openzeppelin/contracts/ownership/Ownable.sol";
+import * as ReentrancyGuard from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import * as SafeMath from "./external/openzeppelin-solidity/math/SafeMath.sol";
+import * as INXMMaster from "./INXMMaster.sol";
+import * as Pool1 from "./Pool1.sol";
+import * as PoolData from "./PoolData.sol";
+import * as Claims from "./Claims.sol";
+import * as ClaimsData from "./ClaimsData.sol";
+import * as NXMToken from "./NXMToken.sol";
+import * as QuotationData from "./QuotationData.sol";
+
+contract Distributor is
+  ERC721.ERC721Full("NXMDistributorNFT", "NXMDNFT"),
+  Ownable.Ownable,
+  ReentrancyGuard.ReentrancyGuard {
+
+  struct TokenData {
+    uint expirationTimestamp;
+    bytes4 coverCurrency;
+    uint[] coverDetails;
+    uint coverId;
+    bool claimInProgress;
+    uint claimId;
+  }
+
+  event PayoutReceived (
+    address sender,
+    uint value,
+    bytes4 currency
+  );
+
+  event ClaimRedeemed (
+    address receiver,
+    uint value,
+    bytes4 currency
+  );
+
+  uint public constant CLAIM_VALIDITY_MAX_DAYS_OVER_COVER_PERIOD = 30 days;
+  uint public constant CLAIM_DEPOSIT_PERCENTAGE = 5;
+
+  INXMMaster.INXMMaster internal nxMaster;
+  uint public priceLoadPercentage;
+  uint256 internal tokenIdCounter;
+  mapping(uint256 => TokenData) internal allTokenData;
+
+  uint public withdrawableETH;
+  uint public withdrawableDAI;
+
+  constructor(address _masterAddress, uint _priceLoadPercentage) public {
+    nxMaster = INXMMaster.INXMMaster(_masterAddress);
+    priceLoadPercentage = _priceLoadPercentage;
+  }
+
+  function buyCover(
+        address coveredContractAddress,
+        bytes4 coverCurrency,
+        uint[] memory coverDetails,
+        uint16 coverPeriod,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+  )
+     public
+     payable
+  {
+    uint requiredValue = priceLoadPercentage.mul(coverDetails[1]).div(100).add(coverDetails[1]);
+    require(msg.value == requiredValue, "Incorrect value sent");
+
+    Pool1.Pool1 p1 = Pool1.Pool1(nxMaster.getLatestAddress("P1"));
+    p1.makeCoverBegin.value(coverDetails[1])(coveredContractAddress, coverCurrency, coverDetails, coverPeriod, _v, _r, _s);
+
+    // add fee to the withdrawable pool
+    withdrawableETH = withdrawableETH.add(requiredValue.sub(coverDetails[1]));
+
+    mintToken(coverCurrency, coverDetails, coverPeriod);
+  }
+
+  function buyCoverUsingCA(
```

Comment:
Can we merge these 2 functions? In terms of interface, I don't see why we'd need 2 functions. Any concerns?